### PR TITLE
Add `SignedQeIdentityVerifier`

### PR DESCRIPTION
**Pull request by nick-mobilecoin**
Add a verifier which verifies the signature and time validity of the qe
identity json data.

Be sure and look at https://github.com/mobilecoinfoundation/attestation/pull/105#discussion_r1232422318 as it may push for a different implementation

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #99** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/99" target="_blank">Graphite</a> 
    * **PR #104** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/104" target="_blank">Graphite</a> 
      * **PR #105** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/105" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/105?utm_source=stack-comment).

File: verifier/src/quote_enclave_identity.rs [hunk 1]
```diff
@@ -107,6 +111,46 @@ pub struct SignedQeIdentity<'a> {
     signature: Vec<u8>,
 }
 
+impl<'a> SignedQeIdentity<'a> {
```
Comment (nick-mobilecoin) [inline verifier/src/quote_enclave_identity.rs:114]: This is very similar to the [TcbInfoRaw](https://github.com/mobilecoinfoundation/attestation/blob/8413e9c404f158241cd854cc381e46586f48ed30/verifier/src/tcb.rs#L234) implementation.
I'm on the fence on how much to worry about the duplication.

One thought I had was:
```rust
#[derive(Debug, Deserialize, Clone)]
pub struct SignedJson<'a> {
    #[serde(borrow, alias="tcbInfo", alias="enclaveIdentity")]
    primary_field: &'a RawValue,
    #[serde(with = "hex")]
    signature: Vec<u8>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SignedJsonTime {
    issue_date: String,
    next_update: String,
}
```

File: verifier/src/quote_enclave_identity.rs [hunk 2]
```diff
@@ -107,6 +111,46 @@ pub struct SignedQeIdentity<'a> {
...
+    }
+
+    fn verify_time(&self, time: DateTime) -> Result<(), Error> {
+        let qe_identity = QeIdentity::try_from(self)?;
+        let issue_date = qe_identity.issue_date.parse::<DateTime>()?;
+        let next_update = qe_identity.next_update.parse::<DateTime>()?;
+        if time < issue_date {
+            Err(Error::QeIdentityNotYetValid)
+        } else if time >= next_update {
+            Err(Error::QeIdentityExpired)
```
Comment (nick-mobilecoin) [inline verifier/src/quote_enclave_identity.rs:147]: I didn't think about it until writing up the SignedJson comment above, but the NotYetValid and Expired are the only errors that are TCB vs QEIdentity.  The error message is "`The QE identity signature could not be verified: {error}`" so especially if the SignedJson approach is more desirable there should be no harm in genericising these errors.

File: verifier/src/quote_enclave_identity.rs [hunk 3]
```diff
@@ -116,11 +160,106 @@ impl<'a> TryFrom<&'a str> for SignedQeIdentity<'a> {
...
+        let mut error = None;
+        if let Err(e) = result {
+            error = Some(e);
+        }
+
+        VerificationOutput::new(error, is_success.into())
+    }
+}
+
+impl VerificationMessage<Option<Error>> for SignedQeIdentityVerifier {
```
Comment (nick-mobilecoin) [inline verifier/src/quote_enclave_identity.rs:210]: IFF the SignedJson approach is desired, I think only these verification messages would need to be implemented for each type.

File: verifier/src/quote_enclave_identity.rs (hunk 1 above)
Comment (eranrund) [reply verifier/src/quote_enclave_identity.rs:114]: I personally wouldn't worry about the duplication, it's a very small amount of simple to follow code (assuming this only repeats for these two types, if we had a bunch more then maybe I would lean towards DRYing).

File: verifier/src/quote_enclave_identity.rs [hunk 4]
```diff
@@ -116,11 +160,106 @@ impl<'a> TryFrom<&'a str> for SignedQeIdentity<'a> {
...
+    type Value = Option<Error>;
+    fn verify(&self, evidence: &E) -> VerificationOutput<Self::Value> {
+        let signed_qe_identity = evidence.get();
+        let result = signed_qe_identity.verify(&self.key, self.time);
+        let is_success = result.is_ok() as u8;
+
+        let mut error = None;
+        if let Err(e) = result {
+            error = Some(e);
+        }
```
Comment (eranrund) [inline verifier/src/quote_enclave_identity.rs:204]: I think this is [`result.err()`](https://doc.rust-lang.org/std/result/enum.Result.html#method.err)

**Review by eranrund — APPROVED** [review]
LGTM, I don't think there's a huge win in going with `SignedJson` and the `alias` trick

File: verifier/src/quote_enclave_identity.rs (hunk 4 above)
Comment (nick-mobilecoin) [reply verifier/src/quote_enclave_identity.rs]: done now

**Pull request merged**